ArduCopter/crash_check.cpp

```diff
@@ -53,6 +58,65 @@ void Copter::crash_check()
     }
 }
 
+// crash_check - disarms motors if a crash has been detected
+// crashes are detected by the vehicle being more than 20 degrees beyond it's angle limits continuously for more than 1 second
+// called at MAIN_LOOP_RATE
+void Copter::thrust_loss_check()
+{
+    static uint16_t thrust_loss_counter;  // number of iterations vehicle may have been crashed
+
+    // exit immediately if motor failure handling already engaged
+    if (motors->get_thrust_boost()) {
+        return;
+    }
+
+    // return immediately if disarmed
+    if (!motors->armed() || ap.land_complete) {
+        thrust_loss_counter = 0;
+        return;
+    }
+
+    // check for desired angle over 15 degrees
+    // todo: add thrust angle to AC_AttitudeControl
+    const Vector3f angle_target =  attitude_control->get_att_target_euler_cd();
+    if ((fabs(angle_target.x) > THRUST_LOSS_CHECK_ANGLE_DEVIATION_CD) || (fabs(angle_target.y) > THRUST_LOSS_CHECK_ANGLE_DEVIATION_CD)) {
+        thrust_loss_counter = 0;
+        return;
+    }
+
+    // check for throttle over 90 % or throttle saturation
+    if ((attitude_control->get_throttle_in() < THRUST_LOSS_CHECK_MINIMUM_THROTTLE) && (!motors->limit.throttle_upper)) {
+        thrust_loss_counter = 0;
+        return;
+    }
+
+    // check for throttle over 90 % or throttle saturation
+    if ((attitude_control->get_throttle_in() < 0.25f)) {
+        thrust_loss_counter = 0;
+        return;
+    }
+
+    // check for decent
+    if (inertial_nav.get_velocity_z() >= 0.0f) {
+        thrust_loss_counter = 0;
+        return;
+    }
+
+    // we may be crashing
+    thrust_loss_counter++;
+
+    // check if thrust loss for 1 second
+    if (thrust_loss_counter >= (THRUST_LOSS_CHECK_TRIGGER_SEC * scheduler.get_loop_rate_hz())) {
+        // log an error in the dataflash
+        Log_Write_Error(ERROR_SUBSYSTEM_THRUST_LOSS_CHECK, ERROR_CODE_FAILSAFE_OCCURRED);
```

Comment:
This isn't exactly a failsafe, but I guess it avoids adding another macro. But there is no message clearing it though.

Reply:
Hmmmm, I thought I put one in but obviously not.

Reply:
OK, I will add the recovery messages in a follow up commit.

Reply:
I think i'm going to have to break my promise to add a recovery message. The problem is that our logic for recovery is down in the motors library and I don't think it would be good to put the recovery message that far down in the code.

I think, if there are no strong objections, we can put this in without the recovery message and see how it goes with testing.